rfcs/20240503-opset-deprecations.md

```diff
@@ -0,0 +1,102 @@
+- [Dynamism RFC P4](https://github.com/openxla/stablehlo/blob/main/rfcs/20230704-dynamism-101.md#p4)
+opset updates: `real_dynamic_slice` vs `dynamic_slice`.
+- Potentially unused ops like stateful `rng` [#597](https://github.com/openxla/stablehlo/issues/597)
+and `map`.
+
+In general (unless the op is unused and can be trivially deleted), the
+deprecation steps will be as follows:
+
+1. Migrate framework uses of redundant ops.
+1. Block serialization of deprecated ops once frameworks migrated.
+1. Migrate uses of the ops to the supported StableHLO op (add builder methods).
+1. Change VHLO legalization to upgrade to the supported op for compatibility.
+1. Remove the redundant StableHLO op.
+1. Remove redundant op from VHLO after 6mo.
+
+## Proposed Opset Changes
+
+### P0: Delete `CreateTokenOp` and `TraceOp`
+
+These ops are both unused as far as we can tell. They can be trivially deleted.
+
+### P1: Deprecate `BroadcastOp`, `DotOp`, `UnaryEinsumOp`
+
+These ops are all a trivial subset of features of another op. I.e. BroadcastOp
+can be represented using BroadcastInDim, DotOp with DotGeneralOp, UnaryEinsum
+with [`einsum` lowering][einsum-lowering].
+These ops will follow the formal deprecation process listed above.
+
+Helper methods can be added to the support op for compatibility, something like:
+`isLhsBroadcast`, `isSimpleDot`, `isUnaryEinsum`.
+
+### P2: Deprecate `RealDynamicSliceOp`, Enhance `DynamicSliceOp`
+
+In terms of naming `stablehlo.dynamic_slice` is more in-model than
+`real_final_v2_dynamic_slice`. However in terms of functionality, per
+[Dynamism RFC P4](https://github.com/openxla/stablehlo/blob/main/rfcs/20230704-dynamism-101.md#p4)
+the behavior of `real_dynamic_slice` is correct. We propose to enhance
+`dynamic_slice_op` to have an identical feature set as `real_dynamic_slice`, and
+deprecate `real_dynamic_slice`. This change will be done with full
+forward and backward compatibility.
+
+One could make the argument that `dot` is a more proper name than `dot_general`,
+and I'm happy to go down that route, but it will likely cause a good deal of
+code churn in community repos. Interested in feedback here.
+
+### P3: Move `CrossReplicaSumOp` to CHLO
+
+The `cross-replica-sum` op  (hyphens not a typo), is just sugar for an
+`all-reduce` op. Even in the XlaBuilder's [xla::CrossReplicaSum][CRS]
+implementation this op is decomposed into an all reduce. We could just remove
+this op, and eventually we may, but we propose to move it to CHLO in the short
+term since frameworks map to this op, and this will keep the refactoring fairly
+trivial.
+
+### P4: (Feedback Requested) Deprecate `MapOp`, `RngOp`, `EinsumOp`, `TorchIndexSelectOp`
+
+These are all ops that seem to have very limited use in StableHLO. It would be
+great to remove them all or move them to CHLO, as opposed to providing long term
+compatibility on ops that aren't needed.
+
+In the interim, we only plan to guarantee the existing 6mo compatibility
```

Comment:
Nit: consider spelling out 6mo. I don't think 6mo is particularly unclear, but 6 months is clearer.